### PR TITLE
SYCL: Implement some dependencies for SYCL 2020 Spec constant design

### PR DESCRIPTION
**Pull request by erichkeane**
The Spec-constant design (https://github.com/intel/llvm/pull/3331) has a
few dependencies on the CFE, including generating unique names for
reachable specialization_id variables and putting the results in the
integration footer.

This patch helps with this effort in 2 ways:
First, it creates a command line option that will be used eventually to
ensure that types with internal linkage have names unique to this
translation unit.  This will eventually be used by the
unique-id/unique-stable-name implementation, but is required as a CC1
option to unblock the Driver implementation.

Second, it generates the integration footer sans the generated names
(which will be added when we have the unique-id/unique-stable-name
implementation in place).  This is necessary to unblock library
implementation of this feature.

### CODE REVIEW

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -4245,6 +4249,22 @@ SYCLIntegrationHeader::SYCLIntegrationHeader(bool _UnnamedLambdaSupport,
                                              Sema &_S)
     : UnnamedLambdaSupport(_UnnamedLambdaSupport), S(_S) {}
 
+void SYCLIntegrationFooter::addVarDecl(VarDecl *VD) {
+  // Step 1: ensure that this is of the correct type(spec-constant template
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:4253]:
```suggestion
  // Step 1: ensure that this is of the correct type-spec-constant template
```

File: clang/lib/Sema/SemaSYCL.cpp [hunk 2]
```diff
@@ -4245,6 +4249,22 @@ SYCLIntegrationHeader::SYCLIntegrationHeader(bool _UnnamedLambdaSupport,
...
+  // specialization).
+  if (!Util::isSyclSpecIdType(VD->getType()))
+    return;
+  // Step 2: ensure that this is a static member, or a namespace-scope.
+  // Note that the isLocalVarDeclorParm excludes thread-local and static-local
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:4258]:
```suggestion
  // Note that isLocalVarDeclOrParm excludes thread-local and static-local
```

File: clang/lib/Sema/SemaSYCL.cpp [hunk 3]
```diff
@@ -111,6 +111,10 @@ class Util {
   /// specialization constant class.
   static bool isSyclSpecConstantType(const QualType &Ty);
 
+  /// Checks whether given clang type is a full specialization of the SYCL
+  /// specialization id class.
+  static bool isSyclSpecIdType(const QualType &Ty);
```
[the remark this answers is not in the document]
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:116]: Thats 2 things I want to fix in this 'class' now :)  I'll do a follow up to fix the others, but I'll do this one 'right' for now.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 4]
```diff
@@ -4304,6 +4356,15 @@ bool Util::isSyclSpecConstantType(const QualType &Ty) {
   return matchQualifiedTypeName(Ty, Scopes);
 }
 
+bool Util::isSyclSpecIdType(QualType Ty) {
+  llvm::StringLiteral Name = "specialization_id";
+  std::array<DeclContextDesc, 3> Scopes = {
+      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "cl"},
+      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "sycl"},
+      Util::DeclContextDesc{Decl::Kind::ClassTemplateSpecialization, Name}};
```
[the remark this answers is not in the document]
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:4364]: I guess none of these need 'clang::'.  I just copy/pasted from above :)  Same as the using specialization_id here directly.  I actually don't have a good reason for it being separate.  I'll do some cleanup in THIS patch for this function, then do the rest in the other patch (#3504) for the rest.

[the remark this answers is not in the document]
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:4364]: So hold off on merging this, i want #3504 to go first. THEN I need to merge with that and do a touch of refactoring of this part.

[the remark this answers is not in the document]
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp]: Ok, #3504 is in now, and has been merged with this.  The changes you requested here have been done.

This patch should be ready for review (and presumably approval :) )?

**Review by mdtoguchi — APPROVED** [review]

**Comment by erichkeane** [comment]
Ping @AaronBallman  or @premanandrao .  This should be ready for review now.

**Review by premanandrao — APPROVED** [review]
LGTM

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Pull request merged**